### PR TITLE
Place and Route Documentation/Internals Refactor

### LINKED ISSUE #25: Move generic geometry functions into root 'geometry' module.

### LINKED ISSUE #26: Move tests inline with code for Place & Route
Match the existing convention of having tests in the same directories as their targets.

### PR DESCRIPTION
**Pull request by mossblaser**
* [x] Move tests inline with code (fixes #26)
* [x] Move generic geometry code into a central place (fixes #25)
* [x] Fix minor PEP-8 related grievances
* [x] Move documentation properly into Sphinx

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -97,6 +101,9 @@
 # If true, keep warnings as "system message" paragraphs in the built documents.
 #keep_warnings = False
 
+# Search Python docs for extra definitions.
+intersphinx_mapping = {'python': ('http://docs.python.org/3.2', None)}
```
Comment (mundya) [inline docs/source/conf.py]: 3.2?  Which we don't test against.

File: docs/source/place_and_route.rst [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+-----------
+
+The key pieces of terminology used are defined below:
+
+Application Graph
+    The `hyper-graph <http://en.wikipedia.org/wiki/Hypergraph>`_ which
+    describes how an application's computational resources (the *vertices*) are
+    connected to each other by *nets*.
+Vertex
+    A *vertex* in *application graph*. Each vertex is mapped onto exactly one
```
Comment (mundya) [inline docs/source/place_and_route.rst]: Tautology?  Also "vertex in *a* application graph"

File: docs/source/place_and_route.rst [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    
+    *Nets* are represented by instances of the :py:class:`rig.netlist.Net`
+    class.
+Resource
+    A *resource* is any finite resource available to a SpiNNaker chip (e.g.
+    SDRAM) which may be consumed by a vertex. *Resources* are allocated to each
+    *vertex* during allocation. Users are welcome to define their own
+    application-specific resources.
+    
+    The type of a *resource* is represented by some unique python
```
Comment (mundya) [inline docs/source/place_and_route.rst]: P for python

File: docs/source/place_and_route.rst [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+Constraint
+    *Constraints* specify additional requirements on how an application graph is
+    placed and routed. For example a constraint might be used to force a
+    particular *vertex* to always be placed on a specific chip.
+    
+    A number of types of *constraint* are defined in
+    :py:class:`rig.place_and_route.constraints`.
+
+.. note::
+    It is worth emphasising that vertices being placed on SpiNNaker *chips*,
```
Comment (mundya) [inline docs/source/place_and_route.rst]: "that vertices *are* placed" maybe?

File: docs/source/place_and_route.rst [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    work and so each vertex will consume a single core of spinnaker chip
+    resource.
+    
+    Vertices which consume no cores are typically only useful when describing
+    external devices connected to the SpiNNaker system.
+    
+    Vertices which consume more than one core are typically only useful when a
+    vertex represents a group of applications which share memory. This is
+    because vertices will always be placed on a single SpiNNaker chip: they
+    cannot be split accross many chips. If an application requires this type of
```
Comment (mundya) [inline docs/source/place_and_route.rst]: sp. across

File: docs/source/place_and_route.rst [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+Common case wrapper
+-------------------
+
+Most applications simply require their application graph be translated into a
+set of data structures describing where binaries need to be loaded and a set of
+routing tables. For most users the :py:func:`rig.place_and_route.wrapper`
+will do exactly this with a minimum of fuss.
+
+.. autofunction:: rig.place_and_route.wrapper
```
Comment (mundya) [inline docs/source/place_and_route.rst]: Sadly the functions which are default arguments get typeset *really* horrifically...

**Comment by mundya** [comment]
Aside from the above; LGTM.

File: docs/source/place_and_route.rst (hunk 2 above)
Comment (mossblaser) [inline docs/source/place_and_route.rst]: Where's the tautology?

File: docs/source/place_and_route.rst (hunk 5 above)
Comment (mossblaser) [inline docs/source/place_and_route.rst]: Ta. Also, joy: vim's spellcheck doesn't check inside `.. note::` ughhh I dislike rst and the universe around it...

File: docs/source/place_and_route.rst (hunk 6 above)
Comment (mossblaser) [inline docs/source/place_and_route.rst]: Yeah. Frustratingly there isn't really a good way around this: either don't use autodoc (and let it get out of sync) or use some sentinel value and swap it out later... :(

To be honest, functions with that many arguments become a massive mess in the docs anyway :/

File: docs/source/place_and_route.rst (hunk 2 above)
Comment (mundya) [inline docs/source/place_and_route.rst]: Er, maybe not.  I thought it odd that vertex was used in a sentence without being defined.

**Pull request merged**